### PR TITLE
GH-461 Improve Mainthread.InvokeOnMainThread

### PR DESCRIPTION
**Pull request by Mrnikbobjeff**
### Description of Change ###
Added a hot path to MainThread.InvokeOnMainThread
- Related to issue #461
### API Changes ###
None
### Behavioral Changes ###
None
### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [ ] Has samples (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

### CODE REVIEW

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Xamarin.Essentials/MainThread/MainThread.shared.cs](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/Xamarin.Essentials/MainThread/MainThread.shared.cs) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Launcher.xml](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/docs/en/Xamarin.Essentials/Launcher.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C8%5C9%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201808091700167810-462%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=4oxifUHESbTi%2F%2Bszf%2BtNcZwDguPxEtas6mrxOG%2BXa8w%3D&st=2018-08-09T16%3A56%3A45Z&se=2018-09-09T17%3A01%3A45Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Xamarin.Essentials/MainThread/MainThread.shared.cs](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/Xamarin.Essentials/MainThread/MainThread.shared.cs) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Launcher.xml](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/docs/en/Xamarin.Essentials/Launcher.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C8%5C9%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201808091700246228-462%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=OPOuFhkr6sFkQiIMFLE%2FDTycEYPm8%2BaTRJbCiGq2RU4%3D&st=2018-08-09T16%3A58%3A27Z&se=2018-09-09T17%3A03%3A27Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: Xamarin.Essentials/MainThread/MainThread.shared.cs [hunk 1]
```diff
@@ -13,6 +13,12 @@ public static partial class MainThread
 
         internal static Task InvokeOnMainThread(Action action)
         {
+            if (IsMainThread)
+            {
+                action();
+                return Task.FromResult(true); // Can not use CompletedTask as .net Standard 1.0 does not support it
```
Comment (jamesmontemagno) [inline Xamarin.Essentials/MainThread/MainThread.shared.cs]: So, can you wrap this in:
```
#if NETSTANDARD1_0
return Task.FromResult(null);
#else
return Task.CompletedTask;
#endif
```

no need to add comments in there.

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Xamarin.Essentials/MainThread/MainThread.shared.cs](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/Xamarin.Essentials/MainThread/MainThread.shared.cs) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Launcher.xml](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/docs/en/Xamarin.Essentials/Launcher.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C8%5C10%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201808101438087254-462%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=1RDAwxaP3%2F0Yrqc%2FsS03q2Xe3pTngGIZb8GpH%2FOyOXU%3D&st=2018-08-10T14%3A34%3A23Z&se=2018-09-10T14%3A39%3A23Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by Mrnikbobjeff** [comment]
Okay, I removed the platform code and added checks to all methods in MainThread.shared

**Comment by dend** [comment]
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[Xamarin.Essentials/MainThread/MainThread.android.cs](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/Xamarin.Essentials/MainThread/MainThread.android.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/MainThread/MainThread.ios.cs](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/Xamarin.Essentials/MainThread/MainThread.ios.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/MainThread/MainThread.shared.cs](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/Xamarin.Essentials/MainThread/MainThread.shared.cs) | :white_check_mark:Succeeded | |
[Xamarin.Essentials/MainThread/MainThread.uwp.cs](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/Xamarin.Essentials/MainThread/MainThread.uwp.cs) | :white_check_mark:Succeeded | |
[docs/en/Xamarin.Essentials/Launcher.xml](https://github.com/Mrnikbobjeff/Essentials/blob/feature/MainThreadImprovements/docs/en/Xamarin.Essentials/Launcher.xml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C8%5C10%5Cf63aadd0-c8f5-13ca-04e7-037e45f8c0b9%5CPullRequest%5C201808101441479591-462%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=tU6207E1J1qtGnCuHhxTKFMIr79w%2F%2Fbqsqmq%2Bghxwzw%3D&st=2018-08-10T14%3A37%3A51Z&se=2018-09-10T14%3A42%3A51Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by jamesmontemagno** [comment]
build

**Pull request merged**